src/NServiceBus.Core/Transports/FileBased/MessagePump.cs

```diff
@@ -0,0 +1,195 @@
+            var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
+            var allTasks = runningReceiveTasks.Values.Concat(new[]
+            {
+                messagePumpTask
+            });
+            var finishedTask = await Task.WhenAny(Task.WhenAll(allTasks), timeoutTask).ConfigureAwait(false);
+
+            if (finishedTask.Equals(timeoutTask))
+            {
+                Logger.Error("The message pump failed to stop with in the time allowed(30s)");
+            }
+
+            concurrencyLimiter.Dispose();
+            runningReceiveTasks.Clear();
+        }
+
+        [DebuggerNonUserCode]
+        async Task ProcessMessages()
+        {
+            try
+            {
+                await InnerProcessMessages().ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // For graceful shutdown purposes
+            }
+            catch (Exception ex)
+            {
+                Logger.Error("File Message pump failed", ex);
+                //await peekCircuitBreaker.Failure(ex).ConfigureAwait(false);
+            }
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                await ProcessMessages().ConfigureAwait(false);
+            }
+        }
+
+        async Task InnerProcessMessages()
+        {
+            while (!cancellationTokenSource.IsCancellationRequested)
+            {
+                var filesFound = false;
+
+                foreach (var filePath in Directory.EnumerateFiles(path, "*.*"))
+                {
+                    filesFound = true;
+
+                    var messageId = Path.GetFileNameWithoutExtension(filePath);
+
+                    var transaction = new DirectoryBasedTransaction(path);
+
+                    transaction.BeginTransaction(filePath);
+
+                    await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                    var task = Task.Run(async () =>
+                    {
+                        try
+                        {
+                            await ProcessFile(transaction, messageId);
+
+                            transaction.Complete();
+                        }
+                        finally
+                        {
+                            concurrencyLimiter.Release();
+                        }
+                    }, cancellationToken);
+
+                    task.ContinueWith(t =>
+                    {
+                        Task toBeRemoved;
+                        runningReceiveTasks.TryRemove(t, out toBeRemoved);
+                    }, TaskContinuationOptions.ExecuteSynchronously)
+                        .Ignore();
+
+                    runningReceiveTasks.AddOrUpdate(task, task, (k, v) => task)
+                        .Ignore();
+                }
+
+                if (!filesFound)
+                {
+                    await Task.Delay(10, cancellationToken);
```

Comment:
We could add FileSystemWatcher to pick up most of new files and increase the active polling delay.